### ISSUE TITLE
Support for half precision arithmetic

### ISSUE THREAD

**Issue by cy69855522** [dmlc/dgl#809]
## 🚀 Feature
Thank you for such a great project. I think supporting semi-precision computing will make this project even better.

## Motivation
Graph data is large, and graph networks are getting deeper and larger. I wanted to be able to use semi-precision computing to reduce GPU memory usage, but I found that DGL didn't seem to support it.

## Additional context
```python
<ipython-input-12-a085616f45fa> in forward(self, data)
     43             g = self.nng(h)
     44             h = h.view(batch_size * n_points, -1)
---> 45             h = self.conv[i](g, h)
     46             h = F.leaky_relu(h, 0.2)
     47             h = h.view(batch_size, n_points, -1)

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    491             result = self._slow_forward(*input, **kwargs)
    492         else:
--> 493             result = self.forward(*input, **kwargs)
    494         for hook in self._forward_hooks.values():
    495             hook_result = hook(self, input, result)

/usr/local/lib/python3.6/dist-packages/dgl/nn/pytorch/conv.py in forward(self, g, h)
    622                 g.update_all(self.message, fn.max('e', 'x'))
    623             else:
--> 624                 g.apply_edges(self.message)
    625                 # Although the official implementation includes a per-edge
    626                 # batch norm within EdgeConv, I choose to replace it with a

/usr/local/lib/python3.6/dist-packages/dgl/graph.py in apply_edges(self, func, edges, inplace)
   2166                                            apply_func=func,
   2167                                            inplace=inplace)
-> 2168             Runtime.run(prog)
   2169 
   2170     def group_apply_edges(self, group_by, func, edges=ALL, inplace=False):

/usr/local/lib/python3.6/dist-packages/dgl/runtime/runtime.py in run(prog)
      9         for exe in prog.execs:
     10             # prog.pprint_exe(exe)
---> 11             exe.run()

/usr/local/lib/python3.6/dist-packages/dgl/runtime/ir/executor.py in run(self)
    202         edge_data = self.fdedge.data
    203         dst_data = self.fddst.data
--> 204         udf_ret = fn_data(src_data, edge_data, dst_data)
    205         self.ret.data = FrameRef(Frame(udf_ret))
    206 

/usr/local/lib/python3.6/dist-packages/dgl/runtime/scheduler.py in _mfunc_wrapper(src_data, edge_data, dst_data)
    930         ebatch = EdgeBatch(graph, (u.data, v.data, eid.data),
    931                            src_data, edge_data, dst_data)
--> 932         return mfunc(ebatch)
    933     _mfunc_wrapper = var.FUNC(_mfunc_wrapper)
    934     msg = ir.EDGE_UDF(_mfunc_wrapper, fdsrc, fdedge, fddst)

/usr/local/lib/python3.6/dist-packages/dgl/nn/pytorch/conv.py in message(self, edges)
    598         """The message computation function.
    599         """
--> 600         theta_x = self.theta(edges.dst['x'] - edges.src['x'])
    601         phi_x = self.phi(edges.src['x'])
    602         return {'e': theta_x + phi_x}

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    491             result = self._slow_forward(*input, **kwargs)
    492         else:
--> 493             result = self.forward(*input, **kwargs)
    494         for hook in self._forward_hooks.values():
    495             hook_result = hook(self, input, result)

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/linear.py in forward(self, input)
     90     @weak_script_method
     91     def forward(self, input):
---> 92         return F.linear(input, self.weight, self.bias)
     93 
     94     def extra_repr(self):

/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py in linear(input, weight, bias)
   1404     if input.dim() == 2 and bias is not None:
   1405         # fused op is marginally faster
-> 1406         ret = torch.addmm(bias, input, weight.t())
   1407     else:
   1408         output = input.matmul(weight.t())

RuntimeError: Expected object of scalar type Half but got scalar type Float for argument #4 'mat1'
```

**Comment by yzh119** [comment]
Currently we have not implemented atomic operations for fp16 yet: https://github.com/dmlc/dgl/blob/master/src/kernel/cuda/atomic.cuh
So it may take us a while to suppot half precision.

As a workaround, you could use the following strategy:
```
import dgl
import dgl.function as fn
g.ndata['x'] = g.ndata['x'].float()
g.update_all(fn.copy('x', 'm'), fn.sum('m', 'y'))
g.ndata['y'] = g.ndata['y'].half()
```

**Comment by yzh119** [comment]
Solved in #2552 .
